File: 26_misc/Cycle/WeightedCycle.go
package main

import (
	"fmt"
	"sort"
)

type WeightedCycle struct {
	n          int
	weightsCw  []int
	weightsCcw []int
	prefixCw   []int
	prefixCcw  []int
}

// NewWeightedCycle 创建一个新的带权重的环
//
//	n: 环上的节点数量
//	weightsCw: 顺时针方向边的权重列表，如果为nil，则默认所有边权重为1. weightsCw[i]表示从i顺时针到(i+1)%n的边权重.
//	weightsCcw: 逆时针方向边的权重列表，如果为nil，则默认所有边权重为1. weightsCcw[i]表示从i逆时针到(i-1+n)%n的边权重.
func NewWeightedCycle(n int, weightsCw, weightsCcw []int) (*WeightedCycle, error) {
	wc := &WeightedCycle{
		n: n,
	}

	if weightsCw == nil {
		wc.weightsCw = make([]int, n)
		for i := range wc.weightsCw {
			wc.weightsCw[i] = 1
		}
	} else if len(weightsCw) == n {
		wc.weightsCw = weightsCw
	} else {
		return nil, fmt.Errorf("顺时针权重数量应为%d，但提供了%d个", n, len(weightsCw))
	}

	if weightsCcw == nil {
		wc.weightsCcw = make([]int, n)
		for i := range wc.weightsCcw {
			wc.weightsCcw[i] = 1
		}
	} else if len(weightsCcw) == n {
		wc.weightsCcw = weightsCcw
	} else {
		return nil, fmt.Errorf("逆时针权重数量应为%d，但提供了%d个", n, len(weightsCcw))
	}

	wc.buildPrefixSums()

	return wc, nil
}

func (wc *WeightedCycle) buildPrefixSums() {
	wc.prefixCw = make([]int, wc.n+1)
	for i := range wc.n {
		wc.prefixCw[i+1] = wc.prefixCw[i] + wc.weightsCw[i]
	}
	wc.prefixCcw = make([]int, wc.n+1)
	for i := range wc.n {
		wc.prefixCcw[i+1] = wc.prefixCcw[i] + wc.weightsCcw[(wc.n-i)%wc.n]
	}
}

func (wc *WeightedCycle) Dist(u, v int) int {
	return min(wc.DistCcw(u, v), wc.DistCw(u, v))
}

// 返回逆时针从from到to的带权距离.
func (wc *WeightedCycle) DistCcw(from, to int) int {
	if from >= to {
		return wc.prefixCcw[from+1] - wc.prefixCcw[to]
	}
	return wc.prefixCcw[from+1] + (wc.prefixCcw[wc.n] - wc.prefixCcw[to])
}

// 返回顺时针从from到to的带权距离.
func (wc *WeightedCycle) DistCw(from, to int) int {
	if to >= from {
		return wc.prefixCw[to] - wc.prefixCw[from]
	}
	return (wc.prefixCw[wc.n] - wc.prefixCw[from]) + wc.prefixCw[to]
}

// 返回环上两点间最短路径的线段表示.
func (wc *WeightedCycle) Segment(u, v int) [][2]int {
	if wc.DistCcw(u, v) <= wc.DistCw(u, v) {
		return wc.SegmentCcw(u, v)
	}
	return wc.SegmentCw(u, v)
}

// 返回逆时针从from到to的路径段.
func (wc *WeightedCycle) SegmentCcw(from, to int) [][2]int {
	if from >= to {
		return [][2]int{{from, to}}
	}
	return [][2]int{{from, 0}, {wc.n - 1, to}}
}

// 返回顺时针从from到to的路径段.
func (wc *WeightedCycle) SegmentCw(from, to int) [][2]int {
	if to >= from {
		return [][2]int{{from, to}}
	}
	return [][2]int{{from, wc.n - 1}, {0, to}}
}

// 返回环上两点间的最短路径.
func (wc *WeightedCycle) Path(u, v int) []int {
	if wc.DistCcw(u, v) <= wc.DistCw(u, v) {
		return wc.PathCcw(u, v)
	}
	return wc.PathCw(u, v)
}

// 返回逆时针从from到to的路径经过的点.
func (wc *WeightedCycle) PathCcw(from, to int) []int {
	var path []int
	if from >= to {
		for i := from; i >= to; i-- {
			path = append(path, i)
		}
	} else {
		for i := from; i >= 0; i-- {
			path = append(path, i)
		}
		for i := wc.n - 1; i >= to; i-- {
			path = append(path, i)
		}
	}
	return path
}

// 返回顺时针从from到to的路径经过的点.
func (wc *WeightedCycle) PathCw(from, to int) []int {
	var path []int
	if to >= from {
		for i := from; i <= to; i++ {
			path = append(path, i)
		}
	} else {
		for i := from; i < wc.n; i++ {
			path = append(path, i)
		}
		for i := 0; i <= to; i++ {
			path = append(path, i)
		}
	}
	return path
}

// 判断x是否在from到to的逆时针路径上.
func (wc *WeightedCycle) OnPathCcw(from, to, x int) bool {
	if x < to {
		x += wc.n
	}
	if from < to {
		from += wc.n
	}
	return to <= x && x <= from
}

// 判断x是否在from到to的顺时针路径上.
func (wc *WeightedCycle) OnPathCw(from, to, x int) bool {
	if from > to {
		to += wc.n
	}
	if from > x {
		x += wc.n
	}
	return from <= x && x <= to
}

// 返回数组a中第一个大于x的元素的索引.
func bisectRight(a []int, x int) int {
	return sort.Search(len(a), func(i int) bool { return a[i] > x })
}

// 逆时针从from出发走特定距离到达的位置.
func (wc *WeightedCycle) JumpCcw(from, distance int) int {
	if distance == 0 {
		return from
	}

	totalWeight := wc.prefixCcw[wc.n]
	distance = distance % totalWeight
	if distance == 0 {
		return from
	}

	target := wc.prefixCcw[wc.n-from-1] + distance
	if target >= totalWeight {
		target -= totalWeight
	}

	pos := bisectRight(wc.prefixCcw, target)
	res := wc.n - (pos - 1)
	if res < wc.n {
		return res
	}
	return 0
}

// 顺时针从from出发走特定距离到达的位置.
func (wc *WeightedCycle) JumpCw(from, distance int) int {
	if distance == 0 {
		return from
	}

	totalWeight := wc.prefixCw[wc.n]
	distance = distance % totalWeight
	if distance == 0 {
		return from
	}

	target := wc.prefixCw[from] + distance
	if target >= totalWeight {
		target -= totalWeight
	}

	pos := bisectRight(wc.prefixCw, target)
	return pos - 1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 26_misc/Cycle/WeightedCycle.py
from typing import List, Tuple, Optional
from bisect import bisect_right


def min2(a: int, b: int) -> int:
    return a if a < b else b


class WeightedCycle:
    """带权重的环类，使用前缀和优化查询，支持有向边权重."""

    __slots__ = "n", "weights_cw", "weights_ccw", "prefix_cw", "prefix_ccw"

    def __init__(
        self,
        n: int,
        weights_cw: Optional[List[int]] = None,
        weights_ccw: Optional[List[int]] = None,
    ):
        """
        - 默认所有边权重为1

        :param n: 环上的节点数量

        :param weights_cw: 顺时针方向边的权重列表，如果提供，长度应为n
                          weights_cw[i]表示从i顺时针到(i+1)%n的边权重

        :param weights_ccw: 逆时针方向边的权重列表，如果提供，长度应为n
                          weights_ccw[i]表示从i逆时针到(i-1)%n的边权重
        """
        self.n = n

        if weights_cw is None:
            self.weights_cw = [1] * n
        elif len(weights_cw) == n:
            self.weights_cw = weights_cw
        else:
            raise ValueError(f"顺时针权重数量应为{n}，但提供了{len(weights_cw)}个")

        if weights_ccw is None:
            self.weights_ccw = [1] * n
        elif len(weights_ccw) == n:
            self.weights_ccw = weights_ccw
        else:
            raise ValueError(f"逆时针权重数量应为{n}，但提供了{len(weights_ccw)}个")

        self._build_prefix_sums()

    def _build_prefix_sums(self):
        self.prefix_cw = [0] * (self.n + 1)
        for i in range(self.n):
            self.prefix_cw[i + 1] = self.prefix_cw[i] + self.weights_cw[i]
        self.prefix_ccw = [0] * (self.n + 1)
        for i in range(self.n):
            self.prefix_ccw[i + 1] = self.prefix_ccw[i] + self.weights_ccw[(self.n - i) % self.n]

    def dist(self, u: int, v: int) -> int:
        return min2(self.dist_ccw(u, v), self.dist_cw(u, v))

    def dist_ccw(self, from_: int, to: int) -> int:
        """逆时针从from_到to的带权距离."""
        if from_ >= to:
            return self.prefix_ccw[from_ + 1] - self.prefix_ccw[to]
        else:
            return self.prefix_ccw[from_ + 1] + (self.prefix_ccw[self.n] - self.prefix_ccw[to])

    def dist_cw(self, from_: int, to: int) -> int:
        """顺时针从from_到to的带权距离."""
        if to >= from_:
            return self.prefix_cw[to] - self.prefix_cw[from_]
        else:
            return (self.prefix_cw[self.n] - self.prefix_cw[from_]) + self.prefix_cw[to]

    def segment(self, u: int, v: int) -> List[Tuple[int, int]]:
        """返回环上两点间最短路径的线段表示"""
        if self.dist_ccw(u, v) <= self.dist_cw(u, v):
            return self.segment_ccw(u, v)
        return self.segment_cw(u, v)

    def segment_ccw(self, from_: int, to: int) -> List[Tuple[int, int]]:
        """逆时针从from_到to的路径段"""
        if from_ >= to:
            return [(from_, to)]
        return [(from_, 0), (self.n - 1, to)]

    def segment_cw(self, from_: int, to: int) -> List[Tuple[int, int]]:
        """顺时针从from_到to的路径段"""
        if to >= from_:
            return [(from_, to)]
        return [(from_, self.n - 1), (0, to)]

    def path(self, u: int, v: int) -> List[int]:
        """返回环上两点间的最短路径"""
        if self.dist_ccw(u, v) <= self.dist_cw(u, v):
            return self.path_ccw(u, v)
        return self.path_cw(u, v)

    def path_ccw(self, from_: int, to: int) -> List[int]:
        """逆时针从from_到to的路径经过的点"""
        if from_ >= to:
            return list(range(from_, to - 1, -1))
        return list(range(from_, -1, -1)) + list(range(self.n - 1, to - 1, -1))

    def path_cw(self, from_: int, to: int) -> List[int]:
        """顺时针从from_到to的路径经过的点"""
        if to >= from_:
            return list(range(from_, to + 1))
        return list(range(from_, self.n)) + list(range(0, to + 1))

    def on_path_ccw(self, from_: int, to: int, x: int) -> bool:
        """x是否在from_到to的逆时针路径上"""
        if x < to:
            x += self.n
        if from_ < to:
            from_ += self.n
        return to <= x <= from_

    def on_path_cw(self, from_: int, to: int, x: int) -> bool:
        """x是否在from_到to的顺时针路径上"""
        if from_ > to:
            to += self.n
        if from_ > x:
            x += self.n
        return from_ <= x <= to

    def jump_ccw(self, from_: int, distance: int) -> int:
        """
        逆时针从from_出发走特定距离到达的位置.
        """
        if distance == 0:
            return from_

        total_weight = self.prefix_ccw[self.n]
        distance %= total_weight
        if distance == 0:
            return from_

        target = self.prefix_ccw[~from_] + distance
        if target >= total_weight:
            target -= total_weight

        pos = bisect_right(self.prefix_ccw, target)
        res = self.n - (pos - 1)
        return res if res < self.n else 0

    def jump_cw(self, from_: int, distance: int) -> int:
        """
        顺时针从from_出发走特定距离到达的位置.
        只有当距离完全达到时才能抵达下一个点.
        """
        if distance == 0:
            return from_

        total_weight = self.prefix_cw[self.n]
        distance %= total_weight
        if distance == 0:
            return from_

        target = self.prefix_cw[from_] + distance
        if target >= total_weight:
            target -= total_weight

        pos = bisect_right(self.prefix_cw, target)
        return pos - 1


if __name__ == "__main__":
    weights_cw = [3, 1, 4, 2, 5]  # 顺时针边权重
    weights_ccw = [2, 3, 1, 5, 4]  # 逆时针边权重
    wc = WeightedCycle(5, weights_cw, weights_ccw)

    # 测试距离计算
    assert wc.dist(0, 2) == 4, "错误: 从0到2的最短距离应为4"
    assert wc.dist_ccw(0, 2) == 11, "错误: 从0到2的逆时针距离应为11"
    assert wc.dist_cw(0, 2) == 4, "错误: 从0到2的顺时针距离应为4"

    # 测试路径查找
    assert wc.path(0, 2) == [0, 1, 2], "错误: 从0到2的路径应为[0, 1, 2]"
    assert wc.segment(0, 2) == [(0, 2)], "错误: 从0到2的路径段应为[(0, 2)]"
    assert wc.path_ccw(0, 2) == [0, 4, 3, 2], "错误: 从0到2的逆时针路径应为[0, 4, 3, 2]"
    assert wc.segment_ccw(0, 2) == [
        (0, 0),
        (4, 2),
    ], "错误: 从0到2的逆时针路径段应为[(0, 0), (4, 2)]"
    assert wc.path_cw(0, 2) == [0, 1, 2], "错误: 从0到2的顺时针路径应为[0, 1, 2]"
    assert wc.segment_cw(0, 2) == [(0, 2)], "错误: 从0到2的顺时针路径段应为[(0, 2)]"
    assert wc.path(2, 0) == [2, 1, 0], "错误: 从2到0的路径应为[2, 1, 0]"
    assert wc.segment(2, 0) == [(2, 0)], "错误: 从2到0的路径段应为[(2, 0)]"

    # 测试路径包含性
    assert not wc.on_path_ccw(0, 2, 1), "错误: 1不在从0到2的逆时针路径上"
    assert wc.on_path_ccw(0, 2, 0), "错误: 0在从0到2的逆时针路径上"
    assert wc.on_path_ccw(0, 2, 2), "错误: 2在从0到2的逆时针路径上"
    assert wc.on_path_ccw(0, 2, 3), "错误: 3在从0到2的逆时针路径上"
    assert wc.on_path_ccw(0, 2, 4), "错误: 4在从0到2的逆时针路径上"
    assert wc.on_path_cw(0, 2, 1), "错误: 1在从0到2的顺时针路径上"

    # 测试跳跃功能
    assert wc.jump_ccw(0, 3) == 4, "错误: 从0出发跳跃3步应到达4"
    assert wc.jump_ccw(0, 11) == 2, "错误: 从0出发跳跃11步应到达2"
    assert wc.jump_ccw(0, 15) == 0, "错误: 从0出发跳跃15步应到达0"
    assert wc.jump_cw(0, 3) == 1, "错误: 从0出发跳跃3步应到达1"
    assert wc.jump_cw(0, 11) == 4, "错误: 从0出发跳跃11步应到达4"
    assert wc.jump_cw(0, 15) == 0, "错误: 从0出发跳跃15步应到达0"
    assert wc.jump_cw(0, 0) == 0, "错误: 从0出发跳跃0步应到达0"
